### PR TITLE
[feature] Auto calculate geo coords or addr

### LINKED ISSUE #44: [ux] Automatically calculate geographic coordinates or address
When the position on the geographic map is updated:

1. if the address field is empty we should automatically get the best address from the geographic coordinates and update the address field with the result
2. if the address field is not empty we should ask the user if the user wants the map to be updated (for now with a simple javascript confirm dialog); If the user replies affirmatively we can do the same described in point 1

When address field is updated:

1. If the geographic coordinates on the map are not set, we can get the geographic coordinates  from the address and update the map accordingly
2 if the geographic coordinates on the map are already set, we should ask the user if the user wants the map to be updated (for now with a simple javascript confirm dialog); If the user replies affirmatively we can do the same described in point 1

For the geocoding operations, we can use [geopy](https://geopy.readthedocs.io/).

We should implement an abstraction that allows to change the provider, trying to support at least Google Maps, Nominatim and ArcGIS in the first version fo this feature.

We should handle failures and retry, retry by default 3 times with a delay of 1 second in case of failures, the geocoding library should have a built-in implementation for this, [see RateLimiter](https://github.com/geopy/geopy/blob/5f8e457b700b59221401726e5983175156b20a0a/geopy/extra/rate_limiter.py#L15).

We will likely need the addition of a few django global settings that allow to set:
- the chosen provider, defaults to `ArcGIS`
- the API key, defaults to `None`
- whether the django app may refuse to start if the testing of the geocoding operation fails, defaults to `True`, eg: `DJANGO_LOCI_STRICT_GEOCODE_TEST`
- amount of retries after failures, default to 3
- delay in seconds before trying again after a failure, default to 1

On startup, the app shall test an arbitrary geocoding operation (eg: get the coordinates of the address of the Python Foundation) and if it fails it will raise an `ImproperlyConfigured` exception if ``DJANGO_LOCI_STRICT_GEOCODE_TEST`` is `True` or it will log an error otherwise (using the python logging facility).

Since this app does not use django-rest-framework, the geocoding operations should be performed via a [simple JSON view API that it is added to the admin as we do in django-netjsonconfig](https://github.com/openwisp/django-netjsonconfig/blob/3d14334391f99e964b85826fe085ea0b8a99b6e6/django_netjsonconfig/base/admin.py#L84-L98) (see `context_view` as a reference), although the logic to perform the geocoding can be stored in a function or a class to improve reusability.

This feature needs to be documented in the README and good tests must be written for it.

### PR DESCRIPTION
**Pull request by NoumbissiValere**
Closes #44
Closes #46

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,5 +2,5 @@ django>=2.2,<3.1
 django-leaflet>=0.23,<0.27
 channels>=2.1.3,<2.5.0
 Pillow>=6.2.0,<=7.1.0
-openwisp-utils>=0.4.0,<0.5.0
```
Comment (NoumbissiValere) [inline requirements.txt]: The error shown by `TravisBuddy` was due to the fact that `openwisp-utils version 0.4.0` was being  installed which doesn't have `openwisp-qa-checks`. 
This is the reason why i had to update it

**Comment by NoumbissiValere** [comment]
@nemesisdesign I think the JavaScript work is ok now

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 2]
```diff
@@ -2,5 +2,5 @@ django>=2.2,<3.1
...
+openwisp-utils~=0.6.0
 geopy>=1.20.0,<1.22.0
```
Comment (nemesisdesign) [inline requirements.txt:6]: geopy current latest version is 2.0, can you make sure we use that and the code works please?

**Review by nemesisdesign — COMMENTED** [review]
Thanks @NoumbissiValere I'm testing it.

In the meanwhile, can you please write a test for these 3 lines to avoid the code coverage to go down? https://coveralls.io/builds/34244754/source?filename=django_loci%2Fapps.py#L33
You could instantiate the appconfig class and run that test singularly or call the ready() method, I think we do something similar in other modules, maybe openwisp-monitoring.

File: requirements.txt [hunk 3]
```diff
@@ -1,5 +1,6 @@
-django>=2.2,<3.2
-django-leaflet>=0.23,<0.28
-channels>=2.3.0,<2.5.0
-Pillow~=7.2.0
+django>=2.2,<3.1
+django-leaflet>=0.23,<0.27
+channels>=2.1.3,<2.5.0
+Pillow>=6.2.0,<=7.1.0
```
Comment (nemesisdesign) [inline requirements.txt:4]: Could you rebase on master so to include this change?
https://github.com/openwisp/django-loci/commit/a8cfe17a8c7d6291dc02f2192fe979cfbd0b13d7

File: django_loci/static/django-loci/js/loci.js [hunk 4]
```diff
@@ -462,22 +464,25 @@ django.jQuery(function ($) {
...
     }
 
-    // drag marker event
-    function updateAddrOnDrag() {
+    // triggers update of the address when the location on the map is changed
+    function updateAddressOnMapChange() {
         var marker = getMarker();
-        marker.on('dragend', function () {
+        if (!marker) { return; }
+        getMap().on('draw:edited', function (e) {
```
Comment (nemesisdesign) [inline django_loci/static/django-loci/js/loci.js:485]: @NoumbissiValere @oltarzewskik :eyes: :point_up: https://leaflet.github.io/Leaflet.draw/docs/leaflet-draw-latest.html#l-draw-event

**Review by nemesisdesign — APPROVED** [review]
Added some further improvements, now is ready, I suggest to test it again on your side to double check.

File: django_loci/static/django-loci/js/loci.js (hunk 4 above)
Comment (NoumbissiValere) [reply django_loci/static/django-loci/js/loci.js:485]: Nice, `edited` is cool. it fires up the callback only the save option is clicked which is perfect. thank you @nemesisdesign 
:+1:

**Comment by NoumbissiValere** [comment]
@nemesisdesign I have tested this work again and it works better. I think it is ready.

**Review by nemesisdesign — APPROVED** [review]
Thanks @NoumbissiValere and thanks also @oltarzewskik for the initial work on this :+1:

**Pull request merged**